Bug: pydantic HttpUrl type causes ValidationError after 2.3.x

### Description

After version 2.3.0, it fails to validate schemas correctly, such as `pytantic.networks.HttpUrl` in pydantic_factory, and throws an exception.

```
E   pydantic.error_wrappers.ValidationError: 1 validation error for `MyModel`
E   url
E     invalid or missing URL scheme (type=value_error.url.scheme)
```

### URL to code causing the issue

_No response_

### MCVE

```python
from typing import Optional
from pydantic import BaseModel, HttpUrl
from polyfactory.factories.pydantic_factory import ModelFactory

class MyModel(BaseModel):
    url: Optional[HttpUrl] = Field(None, title='An image URL')

# ...
class TestFactory(ModelFactory[MyModel]):
    __model__ = MyModel

my_factory = TestFactory.build()
my_instance = my_factory() # <--


```
E   pydantic.error_wrappers.ValidationError: 1 validation error for `MyModel`
E   url
E     invalid or missing URL scheme (type=value_error.url.scheme)
```
```


### Steps to reproduce

```bash
1. Prepare pydantic model that has `HttpUrl` type field.
2. Prepare factory using pydantic_factory.ModelFactory
3. Try to create instance using factory(2)
4. See error
```


### Screenshots

"In the format of: `![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)`"


### Logs

_No response_

### Litestar Version

2.3.0

### Platform

- [X] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

<!-- POLAR PLEDGE BADGE START -->
<a href="https://polar.sh/litestar-org/polyfactory/issues/234">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/polyfactory/issues/234/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/polyfactory/issues/234/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/polyfactory/factories/pydantic_factory.py
+++ b/polyfactory/factories/pydantic_factory.py
@@ -16,7 +16,7 @@ from typing import (
 from polyfactory.exceptions import MissingDependencyException
 from polyfactory.factories.base import BaseFactory
 from polyfactory.field_meta import Constraints, FieldMeta, Null
-from polyfactory.utils.helpers import unwrap_new_type
+from polyfactory.utils.helpers import unwrap_new_type, unwrap_optional
 
 try:
     from pydantic import BaseModel
@@ -166,7 +166,7 @@ class PydanticFieldMeta(FieldMeta):
         )
 
         # pydantic v1 has constraints set for these values, but we generate them using faker
-        if annotation in (AnyUrl, HttpUrl, KafkaDsn, PostgresDsn, RedisDsn, AmqpDsn, AnyHttpUrl):
+        if unwrap_optional(annotation) in (AnyUrl, HttpUrl, KafkaDsn, PostgresDsn, RedisDsn, AmqpDsn, AnyHttpUrl):
             constraints = {}
 
         children: list[FieldMeta] = []

--- a/polyfactory/factories/pydantic_factory.py
+++ b/polyfactory/factories/pydantic_factory.py
@@ -17,6 +17,7 @@ from polyfactory.exceptions import MissingDependencyException
 from polyfactory.factories.base import BaseFactory
 from polyfactory.field_meta import Constraints, FieldMeta, Null
 from polyfactory.utils.helpers import unwrap_new_type, unwrap_optional
+from polyfactory.utils.predicates import is_optional_union
 
 try:
     from pydantic import BaseModel
@@ -84,6 +85,13 @@ class PydanticFieldMeta(FieldMeta):
         name = field_info.alias if field_info.alias and use_alias else field_name
 
         annotation = unwrap_new_type(field_info.annotation)
+
+        if is_optional_union(field_info.annotation):
+            # pydantic v2 do not propagate metadata for Union types #[?]
+            # hence we cannot acquire any constraints w/ straightforward approach
+            from pydantic.fields import FieldInfo
+
+            field_info = FieldInfo.from_annotation(unwrap_optional(annotation))
 
         if metadata := [v for v in field_info.metadata if v is not None]:
             constraints = cls.parse_constraints(metadata=metadata)

--- a/tests/test_data_parsing.py
+++ b/tests/test_data_parsing.py
@@ -1,7 +1,7 @@
 from collections import Counter
 from datetime import date
 from enum import Enum
-from typing import Callable, Literal
+from typing import Callable, Literal, Optional
 
 import pytest
 from pydantic import (
@@ -27,6 +27,7 @@ from pydantic import (
     SecretBytes,
     SecretStr,
 )
+from typing_extensions import TypeAlias
 
 from polyfactory.exceptions import ParameterException
 from polyfactory.factories.pydantic_factory import ModelFactory
@@ -223,3 +224,20 @@ def test_class_parsing() -> None:
 
     with pytest.raises(ParameterException):
         MySecondFactory.build()
+
+
+@pytest.mark.parametrize(
+    "type_",
+    [AnyUrl, HttpUrl, KafkaDsn, PostgresDsn, RedisDsn, AmqpDsn, AnyHttpUrl],
+)
+def test_optional_url_field_parsed_correctly(type_: TypeAlias) -> None:
+    class MyModel(BaseModel):
+        url: Optional[type_]
+
+    class MyFactory(ModelFactory[MyModel]):
+        __model__ = MyModel
+
+    while not (url := MyFactory.build().url):
+        assert not url
+
+    assert MyModel(url=url)  # no validation error raised

